File: server/middleware/localization.js
/**
 * Localization Middleware
 * ========================
 * Detects user locale from Accept-Language header or player profile,
 * and provides locale-specific formatting for currencies, dates, and text.
 *
 * Supported locales: en-US, es-ES, fr-FR, de-DE, pt-BR, ja-JP, ko-KR, zh-CN
 */

// Locale-specific currency symbols and formatting
const CURRENCY_CONFIG = {
    USD: { symbol: '$', position: 'before', decimals: 2 },
    EUR: { symbol: '\u20AC', position: 'before', decimals: 2 },
    GBP: { symbol: '\u00A3', position: 'before', decimals: 2 },
    BRL: { symbol: 'R$', position: 'before', decimals: 2 },
    JPY: { symbol: '\u00A5', position: 'before', decimals: 0 },
    KRW: { symbol: '\u20A9', position: 'before', decimals: 0 },
    CNY: { symbol: '\u00A5', position: 'before', decimals: 2 },
    INR: { symbol: '\u20B9', position: 'before', decimals: 2 }
};

// Locale to default currency mapping
const LOCALE_CURRENCY = {
    'en-US': 'USD',
    'en-GB': 'GBP',
    'es-ES': 'EUR',
    'fr-FR': 'EUR',
    'de-DE': 'EUR',
    'pt-BR': 'BRL',
    'ja-JP': 'JPY',
    'ko-KR': 'KRW',
    'zh-CN': 'CNY',
    'hi-IN': 'INR'
};

// Locale to region mapping (for matchmaking)
const LOCALE_REGION = {
    'en-US': 'us-east',
    'en-GB': 'eu-west',
    'es-ES': 'eu-west',
    'fr-FR': 'eu-west',
    'de-DE': 'eu-central',
    'pt-BR': 'sa-east',
    'ja-JP': 'asia-east',
    'ko-KR': 'asia-east',
    'zh-CN': 'asia-east',
    'hi-IN': 'asia-south'
};

const SUPPORTED_LOCALES = Object.keys(LOCALE_CURRENCY);

/**
 * Express middleware to detect and attach locale info.
 * Sets req.locale, req.currency, req.suggestedRegion
 */
function localeDetector(req, res, next) {
    // Priority: 1) Player profile, 2) Query param, 3) Accept-Language header
    let locale = null;

    // From player profile (if authenticated)
    if (req.player?.locale) {
        locale = req.player.locale;
    }

    // From query parameter
    if (!locale && req.query.locale) {
        locale = req.query.locale;
    }

    // From Accept-Language header
    if (!locale) {
        const acceptLang = req.headers['accept-language'];
        if (acceptLang) {
            locale = _parseAcceptLanguage(acceptLang);
        }
    }

    // Default
    if (!locale || !SUPPORTED_LOCALES.includes(locale)) {
        locale = 'en-US';
    }

    req.locale = locale;
    req.currency = LOCALE_CURRENCY[locale] || 'USD';
    req.suggestedRegion = LOCALE_REGION[locale] || 'us-east';
    req.currencyConfig = CURRENCY_CONFIG[req.currency] || CURRENCY_CONFIG.USD;

    next();
}

/**
 * Parse Accept-Language header and find best match.
 */
function _parseAcceptLanguage(header) {
    const langs = header.split(',')
        .map(part => {
            const [lang, q] = part.trim().split(';q=');
            return { lang: lang.trim(), quality: q ? parseFloat(q) : 1.0 };
        })
        .sort((a, b) => b.quality - a.quality);

    for (const { lang } of langs) {
        // Exact match
        if (SUPPORTED_LOCALES.includes(lang)) return lang;
        // Language-only match (e.g., "en" -> "en-US")
        const prefix = lang.split('-')[0];
        const match = SUPPORTED_LOCALES.find(l => l.startsWith(prefix + '-'));
        if (match) return match;
    }

    return null;
}

/**
 * Format a currency amount for the given locale/currency.
 */
function formatCurrency(amount, currency) {
    const config = CURRENCY_CONFIG[currency] || CURRENCY_CONFIG.USD;
    const formatted = amount.toFixed(config.decimals);
    return config.position === 'before'
        ? `${config.symbol}${formatted}`
        : `${formatted}${config.symbol}`;
}

/**
 * Get locale config for a given locale string.
 */
function getLocaleConfig(locale) {
    return {
        locale,
        currency: LOCALE_CURRENCY[locale] || 'USD',
        region: LOCALE_REGION[locale] || 'us-east',
        currencyConfig: CURRENCY_CONFIG[LOCALE_CURRENCY[locale]] || CURRENCY_CONFIG.USD
    };
}

module.exports = {
    localeDetector,
    formatCurrency,
    getLocaleConfig,
    SUPPORTED_LOCALES,
    LOCALE_CURRENCY,
    LOCALE_REGION,
    CURRENCY_CONFIG
};

File: server-rs/src/middleware/localization.rs
use axum::{
    extract::Request,
    middleware::Next,
    response::Response,
};

use crate::error::AppError;

#[derive(Debug, Clone)]
pub struct LocaleInfo {
    pub locale: String,
    pub currency: String,
    pub currency_symbol: String,
    pub currency_position: String, // "before" or "after"
    pub decimal_places: u32,
    pub region: String,
}

impl Default for LocaleInfo {
    fn default() -> Self {
        Self {
            locale: "en-US".to_string(),
            currency: "USD".to_string(),
            currency_symbol: "$".to_string(),
            currency_position: "before".to_string(),
            decimal_places: 2,
            region: "us-east".to_string(),
        }
    }
}

fn locale_config(locale: &str) -> LocaleInfo {
    match locale {
        "es-ES" => LocaleInfo {
            locale: "es-ES".into(),
            currency: "EUR".into(),
            currency_symbol: "€".into(),
            currency_position: "after".into(),
            decimal_places: 2,
            region: "eu-west".into(),
        },
        "fr-FR" => LocaleInfo {
            locale: "fr-FR".into(),
            currency: "EUR".into(),
            currency_symbol: "€".into(),
            currency_position: "after".into(),
            decimal_places: 2,
            region: "eu-west".into(),
        },
        "de-DE" => LocaleInfo {
            locale: "de-DE".into(),
            currency: "EUR".into(),
            currency_symbol: "€".into(),
            currency_position: "after".into(),
            decimal_places: 2,
            region: "eu-west".into(),
        },
        "pt-BR" => LocaleInfo {
            locale: "pt-BR".into(),
            currency: "BRL".into(),
            currency_symbol: "R$".into(),
            currency_position: "before".into(),
            decimal_places: 2,
            region: "us-east".into(),
        },
        "ja-JP" => LocaleInfo {
            locale: "ja-JP".into(),
            currency: "JPY".into(),
            currency_symbol: "¥".into(),
            currency_position: "before".into(),
            decimal_places: 0,
            region: "asia-east".into(),
        },
        "ko-KR" => LocaleInfo {
            locale: "ko-KR".into(),
            currency: "KRW".into(),
            currency_symbol: "₩".into(),
            currency_position: "before".into(),
            decimal_places: 0,
            region: "asia-east".into(),
        },
        "zh-CN" => LocaleInfo {
            locale: "zh-CN".into(),
            currency: "CNY".into(),
            currency_symbol: "¥".into(),
            currency_position: "before".into(),
            decimal_places: 2,
            region: "asia-east".into(),
        },
        "hi-IN" => LocaleInfo {
            locale: "hi-IN".into(),
            currency: "INR".into(),
            currency_symbol: "₹".into(),
            currency_position: "before".into(),
            decimal_places: 2,
            region: "asia-south".into(),
        },
        _ => LocaleInfo::default(),
    }
}

fn detect_locale(req: &Request) -> String {
    // Check query param first
    if let Some(query) = req.uri().query() {
        for pair in query.split('&') {
            let mut kv = pair.splitn(2, '=');
            if kv.next() == Some("locale") {
                if let Some(val) = kv.next() {
                    return val.to_string();
                }
            }
        }
    }

    // Check Accept-Language header
    if let Some(al) = req.headers().get("accept-language").and_then(|v| v.to_str().ok()) {
        // Parse first language tag
        if let Some(lang) = al.split(',').next() {
            let tag = lang.split(';').next().unwrap_or("en-US").trim();
            return tag.to_string();
        }
    }

    "en-US".to_string()
}

/// Middleware: detects locale and attaches LocaleInfo to request.
pub async fn locale_detector(
    mut req: Request,
    next: Next,
) -> Result<Response, AppError> {
    let locale = detect_locale(&req);
    let info = locale_config(&locale);
    req.extensions_mut().insert(info);
    Ok(next.run(req).await)
}

pub fn format_currency(amount_cents: i64, info: &LocaleInfo) -> String {
    let divisor = if info.decimal_places == 0 { 1.0 } else { 100.0 };
    let val = amount_cents as f64 / divisor;
    let formatted = if info.decimal_places == 0 {
        format!("{}", val as i64)
    } else {
        format!("{:.width$}", val, width = info.decimal_places as usize)
    };
    if info.currency_position == "before" {
        format!("{}{}", info.currency_symbol, formatted)
    } else {
        format!("{} {}", formatted, info.currency_symbol)
    }
}
